### ISSUE TITLE
False positive with variable to field

### ISSUE THREAD

**Issue by ben-manes** [uber/NullAway#98]
In this case, a `@Nullable` field is stored to a local variable. When the field is used in a null condition an error occurs, but if the variable is used then it does not. It seems the data flow analysis does not realize they are the same.

```
error: [NullAway] returning @Nullable expression from method with @NonNull return type
    return (writer == null) ? CacheWriter.disabledWriter() : castedWriter;
    ^
```

```java
<K1 extends K, V1 extends V> CacheWriter<K1, V1> getCacheWriter() {
  @SuppressWarnings("unchecked")
  CacheWriter<K1, V1> castedWriter = (CacheWriter<K1, V1>) writer;
  return (writer == null) ? CacheWriter.disabledWriter() : castedWriter;
}
```

**Comment by msridhar** [comment]
Yeah, NullAway does not track local variable aliases so it misses this case.  Like #35 I think we could support this eventually by doing a more precise analysis in cases where we are going to report an error.  Going to mark this as low priority for now but it would be a good project to take on eventually

**Comment by msridhar** [comment]
BTW thanks for all the reports @ben-manes!  Please keep them coming 😄

**Comment by ben-manes** [comment]
Unfortunately I think that I have to give up on the target project since it is non-idiomatic for performance reasons. It uses a lot of nullable fields that are enabled in certain configurations, as indicated by code generated subclasses. Since those usage paths can only occur when non-null, and NullAway cannot track deeply enough, many false positives are hard to easily resolve without turning the whole thing off. But it was a nice experiment to try this out!

**Comment by msridhar** [comment]
Ok no problem!  One thing is if you control the code generator, you can tweak it to generate [downcasts](https://github.com/uber/NullAway/wiki/Suppressing-Warnings#downcasting) or other [warning suppressions](https://github.com/uber/NullAway/wiki/Suppressing-Warnings) so you can still use NullAway on other more idiomatic parts of the code base.  (The downcasts may not be tolerable if the code is super performance critical.)  We'll keep working on improving NullAway in the meantime.

**Comment by ben-manes** [comment]
Yes, though the codegen is only for minimizing fields and the logic is in the base class. I would expect inlining to remove the downcasts, so it would be free. But the [project](https://github.com/ben-manes/caffeine) has exceeded its complexity budget and this would add obscurity when already challenging. But this is a good trick to know if I decide to introduce NullAway to a work project. Thanks!

**Comment by ben-manes** [comment]
Making progress using your trick, so we'll see how it comes out :)

Another case where this issue occurs in the pattern,

```
boolean create = (foo == null);
if (!create) {
  foo.bar()
}
```

This scenario is about readability of the code by naming the condition when used multiple times in a method. Using the explicit check each time would be okay, but convey less when maintaining and need reparsing of what null means. But due to not tracking the invocation on `foo` is a false positive.

**Comment by msridhar** [comment]
> Making progress using your trick, so we'll see how it comes out :)

Awesome! :)  Regarding your `create` example, yeah, this would be good to handle too.  If you have any rough categorization of how often this case happens vs. the first cast example that could be helpful for prioritization.  The solutions are different: the cast example requires tracking equalities between variables / fields, while the `create` example requires tracking conditional (non-)nullness, facts like "if `create` is false then `foo` is not null.

**Comment by ben-manes** [comment]
In this case it was only once, but that is because I usually extract those out to query methods. Since method tracking isn't supported, that has been a largest case of false positives.

My present goal is to have it pass in order to review the benefits. There are a lot more suppressions than I'd like since inference is too locally scoped. But I also think it may have found one or two possible errors in a JSR adapter, so it will be worthwhile regardless. Since this requires more extensive annotations just like the Checker Framework does, I'm curious to run their analysis afterwards to see how it compares.

**Comment by ben-manes** [comment]
Everything works nicely! I had forgotten why the CheckerFramework is painful, so this is a nice compromise. :)

**Comment by msridhar** [comment]
Great, so glad you got things working!  And thanks again for all the reports.  We'll dig through them more in the new year.

Regarding the query methods, I think we can actually support those more easily with a new annotation, rather than requiring a library model.  I'll re-open another issue specifically around that one.

**Comment by ben-manes** [comment]
Joy, [integrated](https://github.com/ben-manes/caffeine/commit/6226cf916a87435433da9acb5bed737d67fa8ca1) into my project. There are 50 targeted suppressions and neutral impact to code readability. Thanks for all of the hard work :)

**Comment by msridhar** [comment]
> Joy, integrated into my project. There are 50 targeted suppressions and neutral impact to code readability. Thanks for all of the hard work :)

Awesome!  Thanks for all your reports and feedback!

**Comment by wbadam** [comment]
@msridhar, what are the chances of this being picked up? We at Canva are rolling out NullAway in our code base, and there's been pushback from some engineers finding it cumbersome to work around NullAway's limitation. Here's an example that's similar to the `create` example above:

```java
String potentiallyNull = // ...
boolean definitelyNotNull = potentiallyNull != null;

if (definitelyNotNull) {
  int length = potentiallyNull.length();  // <-- NullAway warning
}

if (potentiallyNull != null) {
  int length = potentiallyNull.length();  // <-- No warning
}
```

We'd also be happy to contribute if that's the quickest way forward, just let me know how to help.

**Comment by msridhar** [comment]
@wbadam exciting to hear you are working on rolling out NullAway at Canva!  Given the number of times this has come up, I would be open to trying to address cases like your example.  Unfortunately it may not be a super-quick thing to implement.  An over-simplified view is that right now, NullAway keeps track of a set of variables that are `@Nullable` and a set of variables that are `@NonNull` before and after each line of code in a method.  We would need to change this representation so that we could track that a variable is `@NonNull` _if_ some other condition holds, like some boolean variable is `true`.  I was always hesitant to add such support since the dataflow analysis that computes this information is performance critical for NullAway, and we've always tried to keep the compile-time overhead of NullAway as low as possible, so like other Error Prone checks it can be run on every build.  I think it would take some careful design and measurement to add this extra reasoning without compromising performance too much (it's ok to compromise a bit).  Maybe we could use a flag to control whether the extra reasoning is enabled, but that might make the code excessively ugly, so hopefully we wouldn't need that.

Do you or one of your teammates feel willing / able to dig into this a bit?  Having more help would definitely make the exploration go faster.  I am slammed for at least the next couple of days and won't have time to look more closely.  But I can try to give pointers to the relevant code.

**Comment by wbadam** [comment]
@msridhar certainly happy to look into it, a direction to the relevant code would be much appreciated!

**Comment by msridhar** [comment]
I have renamed this issue to focus on the false positives stemming from cases like those in https://github.com/uber/NullAway/issues/98#issuecomment-354630644.  

> @msridhar certainly happy to look into it, a direction to the relevant code would be much appreciated!

Thanks, @wbadam, I appreciate it.  I still need to do some thinking and exploration as to the best way to support this.  Right now, the state tracked by NullAway during local dataflow analysis is a [`NullnessStore`](https://github.com/uber/NullAway/blob/master/nullaway/src/main/java/com/uber/nullaway/dataflow/NullnessStore.java), which maps access paths to their nullness state.  To support this use case we need to instead track something like "conditional nullness", i.e., facts of the form, "if this variable is true/false, then this access path is nullable / non-null."  This would involve updating the store data structure to be able to hold such conditional nullness facts (and to properly compute [least upper bounds](https://github.com/uber/NullAway/blob/master/nullaway/src/main/java/com/uber/nullaway/dataflow/NullnessStore.java#L167-L167) on the stores), updating the transfer functions in [`AccessPathNullnessPropagation`](https://github.com/uber/NullAway/blob/master/nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessPropagation.java) to generate the facts, and also updating relevant APIs that query the dataflow analysis for nullness information at different program points.  And, I would like to do this in a way that does not hurt NullAway performance too much in cases where conditional nullness facts do not need to be tracked.

I have some other deadlines and an upcoming holiday, so I won't be able to dig into this more until after mid-April.  If anyone has cycles and wants to attempt some prototyping based on the above, feel free.  Otherwise, I will take a look when I have some time available.

**Comment by cpovirk** [comment]
Our C++ colleagues at Google have done some work to handle variables like the `definitelyNotNull` example above. My understanding is that they're currently using satisfiability checking but that they are considering alternative approaches that might perform better, including essentially "inlining" `potentiallyNull != null` in place of usages of `definitelyNotNull` (potentially inlining further recursively). (I'm sure there would be caveats around cases in which `potentiallyNull` might change value, for example.) I'm told that one approach (similar but not identical to "inlining") is being demoed in https://github.com/llvm/llvm-project/pull/82950.

**Comment by msridhar** [comment]
Thanks a lot @cpovirk!  That's very interesting.  The inlining approach is appealing in terms of trying to make the results of dataflow not depend on whether you store certain types of conditions in an (effectively final?) local variable or check them directly.  I'm still unsure of the best way to proceed for NullAway.  I hope to be able to put in more time on this soon.

**Comment by agrieve** [comment]
Hitting this a reasonable amount as well for Chrome code.  [Same examples](https://source.chromium.org/chromium/chromium/src/+/main:base/android/java/src/org/chromium/base/BuildInfo.java;drc=f0b563906634a80021db92ce55d681bbb4b42e86;l=274) as above. E.g.:

```
        boolean hostInformationProvided =
                providedHostPackageName != null
                        && providedHostPackageLabel != null
                        && providedHostVersionCode != null
                        && providedPackageName != null
                        && providedPackageVersionName != null;
     if (hostInformationProvided) { ... }
```

**Comment by msridhar** [comment]
I've been meaning to look more at this for a while, just haven't had the cycles.  I'll see if I can prioritize it.  Thank you @agrieve for the additional example.

**Comment by dyrpsf** [comment]
Hi all! I've been looking into this and just opened a Draft PR (#1596) that prototypes a potential architectural fix.

I've extended NullnessStore to track conditionalIfTrue and conditionalIfFalse facts when boolean variables are assigned conditions. I opened it as a draft to get some early feedback on the data structure changes before I wire up the heavy CFG extraction logic to make the test pass. Would love to hear your thoughts on this approach!
